File: Cpp/Valid-Palindrome.cpp
#include <string>

class Solution {
public:
    /**
     * @param s A string
     * @return Whether the string is a valid palindrome
     */
    bool isalphanumeric(char ch) {
        if ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z')) return true;
        else if (ch >= '0' && ch <= '9') return true;
        else return false;
    }
    bool isequal(char ch1, char ch2) {
        if (ch1 >= 'A' && ch1 <= 'Z') ch1 = 'a' + ch1 - 'A';
        if (ch2 >= 'A' && ch2 <= 'Z') ch2 = 'a' + ch2 - 'A';
        if (ch1 == ch2) return true;
        else return false;
    }
    bool isPalindrome(string& s) {
        // Write your code here
        int left = 0, right = s.length() - 1;
        while (left < right) {
            while (left < right && !isalphanumeric(s[left])) left ++;
            while (left < right && !isalphanumeric(s[right])) right --;
            if (left < right && !isequal(s[left], s[right])) {
                return false;
            } else {
                left ++;
                right --;
            }
        }
        return true;
    }
};

File: Java/Valid-Palindrome.java
public class Solution {
    /**
     * @param s A string
     * @return Whether the string is a valid palindrome
     */
    public boolean isPalindrome(String s) {
        // Write your code here
        int left = 0, right = s.length() - 1;
        while (left < right) {
            while (left < right && !Character.isAlphabetic(s.charAt(left)) && !Character.isDigit(s.charAt(left))) left ++;
            while (left < right && !Character.isAlphabetic(s.charAt(right)) && !Character.isDigit(s.charAt(right))) right --;
            if (left < right) {
                char ch1, ch2;
                if (Character.isUpperCase(s.charAt(left))) ch1 = Character.toLowerCase(s.charAt(left));
                else ch1 = s.charAt(left);
                if (Character.isUpperCase(s.charAt(right))) ch2 = Character.toLowerCase(s.charAt(right));
                else ch2 = s.charAt(right);
                if (ch1 != ch2) return false;
                left ++;
                right --;
            }
        }
        return true;
    }
}